feat(py_console_script_binary)!: entry points with custom dependencies

aignas opened the pull request:
Add `py_console_script_binary`, a macro/rule that allows better customization of how
entry points are generated. Notable features of it are:
  * It allows passing in additional dependencies, which makes it easier for plugin
    dependencies to be added to tools such as pylint or sphinx.
  * The underlying `py_binary` rule can be passed in, allowing custom rules,
    such as the version-aware rules, to be used for the resulting binary.
  * Entry point generation is based upon a wheel's `entry_points.txt` file. This helps
    avoid loading external repositories unless they're actually used, allows entry
    points to have better version-aware support, and allows bzlmod to provide a
    supportable mechanism for entry points.

Because the expected common use case is an entry point for our pip generated repos,
there is special logic to make that easy and concisely do. Usage of
`py_console_script_binary` is not tied to our pip code generation, though, and users can
manually specify dependencies if they need to.

BREAKING CHANGE: This is a breaking change, but only for bzlmod users. Note that
bzlmod support is still beta. Bzlmod users will need to replace using `entry_point`
from `requirements.bzl` with loading `py_console_script_binary` and defining the
entry point locally:

```
load("@rules_python//python/entry_points:py_console_script_binary.bzl, "py_console_script_binary")

py_console_script_binary(name="foo", pkg="@mypip//pylint")
```

For workspace users, this new macro is available to be used, but the old code is still
present.

Fixes #1362
Fixes #543
Fixes #979
Fixes #1262
Closes #980
Closes #1294
Closes #1055

aignas commented:
Ideally, I would love to reuse the implementation for `py_binary` and have `entry_point` as an actual rule, but genrule can also work here initially.

@rickeylev, @groodt, this may have some rough edges still (I am wondering if the workaround around the `sys.path` entries yields to correct behaviour, but I am not sure if there is a better solution yet), but it is ready for initial feedback. The benefit of this approach is that it does not require very clever tricks explored in #1294 in order to expose spoke contents via the hub repo and we can just use the stuff exposed in the hub repo directly.

chrislovecnm commented:
@aignas I'm getting the same error on Windows.  Not sure if this helps:

```
PS C:\Users\chris\Workspace\rules_python\examples\bzlmod\bazel-bin\entry_point> c:\\b\\r4du22do\\execroot\\_main\\bazel-out\\x64_windows-fastbuild\\bin\\entry_point\\entry_point_test.exe.runfiles\\_main\\entry_point\\yamllint.exe -v
Traceback (most recent call last):
  File "C:\Users\chris\AppData\Local\Temp\Bazel.runfiles_mbclvpp2\runfiles\_main\entry_point\yamllint.py", line 12, in <module>
    from yamllint.cli import run
  File "C:\Users\chris\AppData\Local\Temp\Bazel.runfiles_mbclvpp2\runfiles\_main\entry_point\yamllint.py", line 12, in <module>
    from yamllint.cli import run
ModuleNotFoundError: No module named 'yamllint.cli'; 'yamllint' is not a package
PS C:\Users\chris\Workspace\rules_python\examples\bzlmod\bazel-bin\entry_point
```

rickeylev commented:
FYI: I have some pubic API-surface feedback to give, but gtg at the moment.

Skimming the PR, I'll also likely have some implementation feedback, but need to look more closely.

chrislovecnm commented on examples/bzlmod/entry_point/BUILD.bazel:
```diff
@@ -1,18 +1,38 @@
+load("@rules_python//python:py_entry_point_binary.bzl", "py_entry_point_binary")
 
-alias(
+# We can specifically request the py_entry_point_binary to be running with Python 3.9.
```
Please include more documentation of what we are supporting here.  What is an entry_point, and how can a user use this?

chrislovecnm commented on examples/bzlmod/entry_point/BUILD.bazel:
```diff
@@ -1,18 +1,38 @@
+    # the package.
+)
+
+# Here we are running the default version.
```
Please explain what a "default" version is.  Approach the documentation from a user that is new to bazel.

chrislovecnm commented on examples/bzlmod/entry_point/test_entry_point.py:
```diff
@@ -38,6 +39,58 @@ def test_entry_point(self):
         )
         self.assertEqual(proc.stdout.decode("utf-8").strip(), "yamllint 1.28.0")
 
+    def test_pylint_entry_point(self):
```
Should we document what the test is doing and how it works?

chrislovecnm commented on examples/bzlmod/entry_point/test_entry_point.py:
```diff
@@ -38,6 +39,58 @@ def test_entry_point(self):
+
+    def test_pylint_entry_point_deps(self):
+        rlocation_path = os.environ.get("PYLINT_ENTRY_POINT")
+        entry_point = pathlib.Path(runfiles.Create().Rlocation(rlocation_path))
```
Should we run `self.assertTrue(entry_point.exists(), f"'{entry_point}' does not exist")`

chrislovecnm commented on examples/pip_parse_vendored/BUILD.bazel:
```diff
@@ -16,7 +16,7 @@ genrule(
     cmd = " | ".join([
```
So I hate to ask this, but can we do this with starlark and not bash?  A lot of windows users don't have bash installed.

aignas replied:
No, please see #979 for reasons why.

aignas replied:
I'd like to leave this out of scope of this PR.

chrislovecnm commented on examples/pip_parse_vendored/requirements.bzl:
```diff
@@ -35,6 +35,21 @@ def dist_info_requirement(name):
+
+    # buildifier: disable=print
+    print("""
+DEPRECATED: please replace this instance of entry_point with the following:
```
I love that you put in a depreciation notice! Can we include a link to the docs?

aignas replied:
Will do once we have agreed on design.

chrislovecnm commented on python/pip_install/pip_repository_requirements.bzl.tmpl:
```diff
@@ -34,6 +34,21 @@ def dist_info_requirement(name):
+
+    # buildifier: disable=print
+    print("""
+DEPRECATED: please replace this instance of entry_point with the following:
```
Please include a link to the docs, and mention that you must put this rule into a project's BUILD file.

chrislovecnm commented on python/pip_install/tools/entry_point_generator/generator.py:
```diff
@@ -0,0 +1,144 @@
+        raise RuntimeError(
+            f"The package does not provide any console_scripts in it's {_ENTRY_POINTS_TXT}"
+        )
+        return 1
```
Dumb question.  If you raise an error, will you get to this line?

chrislovecnm commented on python/pip_install/tools/entry_point_generator/generator.py:
```diff
@@ -0,0 +1,144 @@
+    )
+    parser.add_argument(
+        "--script",
+        help="The script to generate the entry_point template for.",
```
Please add the "required" parameter on the args that are "required."

```python
     required= True,
```

aignas replied:
This is not required if there is only a single `console_script` in the `entry_points.txt` file.

chrislovecnm commented on python/pip_install/tools/entry_point_generator/generator.py:
```diff
@@ -0,0 +1,144 @@
+        help="The script to generate the entry_point template for.",
+    )
+    parser.add_argument(
+        "entry_points",
```
You have `--shebang` with `--` prefix, and the next two arguments you do not have `--` prefix.  Thoughts?

aignas replied:
These are positional arguments and that is `argparse` specifics. Do you want to make them non-positional?

chrislovecnm commented on python/pip_install/tools/entry_point_generator/generator.py:
```diff
@@ -0,0 +1,144 @@
+        "out",
+        type=pathlib.Path,
+        metavar="OUT",
+        help="The output file.",
```
output file or what?  I assume it is the entry point script.

chrislovecnm commented on python/private/py_entry_point_gen.bzl:
```diff
@@ -0,0 +1,88 @@
+            mandatory = True,
+        ),
+        "out": attr.output(
+            doc = "Output file location.",
```
What are we outputting?

chrislovecnm commented on python/private/py_entry_point_gen.bzl:
```diff
@@ -0,0 +1,88 @@
+            mandatory = True,
+        ),
+        "script": attr.string(
+            doc = "The name of the script to create the entry_point file for.",
```
Not sure what you are trying to say here.

chrislovecnm commented on python/private/toolchains_repo.bzl:
```diff
@@ -194,6 +202,13 @@ def py_binary(name, **kwargs):
         **kwargs
     )
 
+def py_entry_point_binary(name, **kwargs):
```
Should we add docs here?  Any reason why we are using **kwargs?  I am not a huge fan of it :(

aignas replied:
We shouldn't do it here IMO, because `py_binary` has too many of them and I would not like to duplicate docs for them.

chrislovecnm commented on python/py_entry_point_binary.bzl:
```diff
@@ -0,0 +1,92 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
```
Love the docs!! Whoop

chrislovecnm commented on python/py_entry_point_binary.bzl:
```diff
@@ -0,0 +1,92 @@
+        binary_rule: The binary rule to call to create the entry_point binary.
+            Defaults to @rules_python//python:py_binary.bzl#py_binary.
+        deps: The extra dependencies to add to the binary_rule rule.
+        **kwargs: Extra parameters forwarded to binary_rule.
```
Do we need `**kwargs`?  If so can you provide examples when we need it?

aignas replied:
There is an example for `deps` in the `bzlmod/entry_point`, I hope it is sufficient.

chrislovecnm commented on python/py_entry_point_binary.bzl:
```diff
@@ -0,0 +1,92 @@
+    """
+    main = "rules_python_entry_point_{}.py".format(name)
+
+    # This may come via transitions, so ensure that we are not using it at all.
```
What do you mean?

chrislovecnm reviewed (changes requested):
Overall looks good! I know that @rickeylev has some input as well.  Mostly my comments are about documentation.  I want to include more documentation to simplify code maintenance and our users' lives.

@rickeylev and I have talked about how much we appreciate all of the work you have done!

chrislovecnm commented:
@aignas are we fixing https://github.com/bazelbuild/rules_python/issues/1055 as well?

aignas commented:
I'll be going on vacation from tomorrow, so will not respond to questions for a week and a bit.

rickeylev commented on python/entry_point/generator.py:
```diff
@@ -0,0 +1,144 @@
+    elif not console_script:
+        console_script = list(console_scripts.keys())[0]
+
+    entry_point = console_scripts[console_script]
```
```suggestion
    try:
        entry_point = console_scripts[console_script]
    except KeyError as e:
        raise RuntimeError(f"Console script {console_script} not found") from None
```

rickeylev commented on python/entry_point/generator.py:
```diff
@@ -0,0 +1,144 @@
+        )
+        return 1
+
+    if len(console_scripts) > 1 and not console_script:
```
IMHO, this logic would be a bit cleaner if written in a "console script first" manner, but it's up to you.

```
if len(console_scripts) == 0:
  fail("no console scripts")
elif console_script:
  entry_point = console_scripts[console_script] or fail('{console script} not found, {keys} available')
elif len(console_scripts) == 1:
  entry_point =  console_script keys [0]
else:
  fail("must specify console script, available are {keys}")
```

rickeylev commented on python/entry_point/generator.py:
```diff
@@ -0,0 +1,144 @@
+
+    entry_point = console_scripts[console_script]
+
+    module, _, entry_point = entry_point.rpartition(":")
```
Entry points can also have "extras":

```
[console_scripts]
foo = foomod:main
# One which depends on extras:
foobar = foomod:main_bar [bar,baz]
```

I'm fine with leaving a TODO to address this later.

rickeylev commented on python/entry_point/generator_test.py:
```diff
@@ -0,0 +1,190 @@
+#!/usr/bin/env python3
```
Omit the shebang, it's ignored

rickeylev commented on python/entry_point/generator_test.py:
```diff
@@ -0,0 +1,190 @@
+            entry_points = tmpdir / "entry_points.txt"
+            entry_points.write_text(given_contents)
+
+            with self.assertRaises(RuntimeError) as cm:
```
When asserting a generic exception, check the error message to verify it looks like what you expect. This prevents a bug from triggering the same exception and the testing thinking it caught the intended error, when it should have failed.

aignas replied:
I think this is what I am doing a few lines below, where I am using the `cm` and checking the args of the `RuntimeError`.

rickeylev commented on python/entry_point/generator_test.py:
```diff
@@ -0,0 +1,190 @@
+            with self.assertRaises(RuntimeError) as cm:
+                run(
+                    entry_points=entry_points,
+                    out=pathlib.Path(__file__),
```
Isn't this making the test overwrite itself with the entry point output...?? Create a separate file for the output instead, please.

Alternatively, refactor the code to allow passing an io.StringIO object. This avoids much of the boilerplate to setup temp directories.

rickeylev commented on python/entry_point/generator_test.py:
```diff
@@ -0,0 +1,190 @@
+
+            got = out.read_text()
+
+        want = textwrap.dedent(
```
Asserting the exact content for one test case is fine; it lets one see what the generated code is supposed to look like a glance.

For variations of the base case, though, just search for the substrings of interest.

Otherwise, if the template is changed, we have to update more of the tests unnecessarily.

rickeylev commented on python/pip_install/pip_repository_requirements_bzlmod.bzl.tmpl:
````diff
@@ -30,4 +30,19 @@ def entry_point(pkg, script = None):
+    fail("""Please replace this instance of entry_point with the following:
+
+```
+load("@rules_python//python/entry_point:py_console_script_binary.bzl", "entry_point")
````
```suggestion
load("@rules_python//python/entry_point:py_console_script_binary.bzl", "py_console_script_binary")
```

rickeylev commented on python/pip_install/pip_repository_requirements_bzlmod.bzl.tmpl:
````diff
@@ -30,4 +30,19 @@ def entry_point(pkg, script = None):
+```
+load("@rules_python//python/entry_point:py_console_script_binary.bzl", "entry_point")
+
+entry_point(
````
```suggestion
py_console_script_binary(
```

rickeylev commented on python/entry_point/generator.py:
Please move to a private directory. The targets and things in //python/entry_points should just be a our public API. All the internals should be somewhere under //python/private (some of those may be marked visibility=public due to private attribute dependency relationships, but thats OK).

rickeylev reviewed (changes requested):
Ah, I see you've gone through the trouble of implementing the "read from entry points config file" approach I originally was thinking to defer on :).

Well, ok, but lets we need to put some more thought into the design and better document what parts of an entry point config we do and don't support. Lets have that discussion on #979 instead of this PR. I'll post further thoughts on it there. I've tried to refrain from over-commenting on things relating to that.

aignas commented on python/entry_point/py_console_script_binary.bzl:
```diff
@@ -0,0 +1,87 @@
+    """
+    main = "rules_python_entry_point_{}.py".format(name)
+
+    # This may come via transitions, so ensure that we are not using it at all.
```
A leftover from the past, removed.

aignas commented on python/entry_point/private/py_entry_point_gen.bzl:
```diff
@@ -0,0 +1,88 @@
+    attrs = {
+        "console_script": attr.string(
+            doc = "The name of the console_script to create the .py file for.",
+            default = "",
```
Updated the docs to clearly indicate when it is optional.

aignas commented on python/entry_point/private/py_entry_point_gen.bzl:
```diff
@@ -0,0 +1,88 @@
+_TOOLCHAIN_TYPE = "//python:toolchain_type"
+
+def _get_entry_points_txt(dist_info):
+    for file in dist_info.files.to_list():
```
Usually a handful (less than 10), but I agree that it may be better to do this differently. I'll leave this for now as is.

rickeylev replied:
SGTM

aignas commented on python/entry_point/generator.py:
```diff
@@ -0,0 +1,144 @@
+# break due to this reason (e.g. pylint). This means that we should not use this
+# script to ever generate entry points for scripts within the main workspace,
+# but that is fine, we can create a separate generator or a boolean flag for
+# that.
```
I dug a little deeper and I am confused. The `pylint` fails to start when running via `bazel run` but succeeds when running within the test. I am sure that `rules_python` users would use `pylint` `entry_point` via both mechanisms but I am not sure what exactly is going on and it is getting late.

What I checked:
* The `sys.path` is the same in both instances except for the additional entry to load the `@rules_python//python/runfiles`.
* The entries for the test are using the `sandox` folders whereas `bazel run` is using something else:
  * See `bazel run` syspath here: https://gist.github.com/aignas/396ed948a9a51a59bdb6c4155347c3fe
  * See `bazel test` syspath here: https://gist.github.com/aignas/f909306720d430bcd917444a9791fdc0

And I am not sure why `pylint` starts working if I remove the first `sys.path` entry, possibly because the directory contains the following:
```
.r-xr-xr-x  21k aignas 21 Aug 00:13 entry_point_test
.r-xr-xr-x  34k aignas 21 Aug 00:13 entry_point_test.repo_mapping
drwxr-xr-x    - aignas 21 Aug 00:13 entry_point_test.runfiles
.r-xr-xr-x 1.8M aignas 21 Aug 00:13 entry_point_test.runfiles_manifest
.r-xr-xr-x  21k aignas 21 Aug 00:26 pylint
.r-xr-xr-x  14k aignas 21 Aug 00:26 pylint.repo_mapping
drwxrwxrwx    - aignas 21 Aug 00:26 pylint.runfiles
.r-xr-xr-x 1.6M aignas 21 Aug 00:26 pylint.runfiles_manifest
.r-xr-xr-x  21k aignas 21 Aug 00:26 pylint_deps_test
.r-xr-xr-x  20k aignas 21 Aug 00:26 pylint_deps_test.repo_mapping
drwxr-xr-x    - aignas 21 Aug 00:26 pylint_deps_test.runfiles
.r-xr-xr-x 1.6M aignas 21 Aug 00:26 pylint_deps_test.runfiles_manifest
.r-xr-xr-x  21k aignas 21 Aug 00:26 pylint_test
.r-xr-xr-x  16k aignas 21 Aug 00:26 pylint_test.repo_mapping
drwxr-xr-x    - aignas 21 Aug 00:26 pylint_test.runfiles
.r-xr-xr-x 1.6M aignas 21 Aug 00:26 pylint_test.runfiles_manifest
.r-xr-xr-x  21k aignas 21 Aug 00:26 pylint_with_deps
.r-xr-xr-x  18k aignas 21 Aug 00:26 pylint_with_deps.repo_mapping
drwxr-xr-x    - aignas 21 Aug 00:26 pylint_with_deps.runfiles
.r-xr-xr-x 1.6M aignas 21 Aug 00:26 pylint_with_deps.runfiles_manifest
.r-xr-xr-x  619 aignas 21 Aug 00:28 rules_python_entry_point_pylint.py
.r-xr-xr-x  619 aignas 21 Aug 00:28 rules_python_entry_point_pylint_with_deps.py
.r-xr-xr-x  611 aignas 21 Aug 00:28 rules_python_entry_point_yamllint.py
.r-xr-xr-x  21k aignas 21 Aug 00:26 yamllint_test
.r-xr-xr-x 3.4k aignas 21 Aug 00:26 yamllint_test.repo_mapping
drwxr-xr-x    - aignas 21 Aug 00:26 yamllint_test.runfiles
.r-xr-xr-x 1.6M aignas 21 Aug 00:26 yamllint_test.runfiles_manifest
```

Hence I am thinking that actually removing the first entry would yield the correct behaviour so that we would not be able to `import` from outside the `runfiles` directory.

aignas replied:
I have put the workaround as I am somewhat 90% convinced that it is needed but would like to have a better mental model of what the differences are between `bazel run` and `bazel test` in this situation. I have also attempted to create an `alias` and name the underlying `py_binary` something else, but `pylint` was still failing if the first item in the `sys.path` was present.

rickeylev replied:
The `bazel run` paths look like what a non-sandboxed test execution would look like which is basically what it is. So that makes sense. Disabling sandbox or forcing local execution would probably make the test look more similar.

Looking at your pastes, I found a couple apparent typos, and 1 potentially meaningful difference:

* The test has 1 extra entry: `pylint_with_deps/runfiles/rules_python~override`, i.e. the bzlmod module root. Is there a directory in there that might be confusing things? From what I recall, all the directories in there have "bzlmod canonical" names (`bla~whatever~blabla`), but don't 100% recall
* Typo-looking things (run vs test):
  * `pip_39_` vs `pip_3_`
  * `python_3_9_x86_64` vs `pyhon_3_9_x86_64`

This all very much sounds like the "sys.path[0] has the script dir" issue that PYTHONSAFEPATH was created to address.

Does this go away when using Bazel at head with Python 3.11? (e.g. USE_BAZEL_VERSION=last_green). That combination should prevent sys.path[0] from being the main py dir.

It looks like sys.flags.isolated and sys.flags.safe_path could be checked to see if the behavior is available/enabled.

Alternatively, try 3.11, set PYTHONSAFEPATH=1 in the env, then re-exec sys.argv and see what sys.path looks like.

Anyways, I think the sys.path[0] removal you have is fine enough.

aignas replied:
So:
1. I removed the sys.path removal workaround
2. I added pylint with requirements for Python 3.11
3. And then I created a separate target to force usage of Python 3.11
4. Ran `USE_BAZEL_VERSION=last_green bazel run //entry_points:pylint_latest` and it just worked as expected, whereas the `Python 3.9` target was still failing. So I think your analysis is correct here.

Diff used to experiment in the gist: https://gist.github.com/aignas/e5097e93bd7117d95dfe1e557865784a

That lead me to add a line to check about `safe_path`, but I left out `sys.flags.isolated` checks because it was always `0` in my local testing, so we can leave this for a followup if really needed.

aignas commented:
I'd be grateful if somebody could enlighten me as to why the Windows tests are failing.

rickeylev commented:
> I'd be grateful if somebody could enlighten me as to why the Windows tests are failing.

Given the error, it looks like the entry point's executable (the "windows launcher" wrapper thing) is running, and that the launcher is failing to find...something. There's a lot of something it could be :(. Do you have a windows machine?

The basic way the windows+zip mechanism work is...

A native windows executable is created. This is called the "windows launcher". This binary has the path to python interpreter to use embedded into it by the py_binary build process. That interpreter path is uses comes from...it's supposed to be the toolchain. However, I seem to recall that `--python_path` could override it (and might have on windows?), but I'm having trouble finding the logic that did that.

Additionally, it looks like, in the launcher source code, it does `if <desired path doesn't exist>: use python.exe`.

A few ideas:

1. Strip down the test to just call the entry point without any args. That eliminates any of the other args being passed as potential issues.
2. Try running `subprocess.run("python.exe", shell=True)`. It's not clear to me if it's invoking python.exe via a shell (which would trigger PATH searching), or directly (e.g. via exec, which wouldn't perform PATH searching. I have to assume the former; I don't see how it could have worked without PATH lookup.
3. The launcher source seems to indicate it has a special `--print_launcher_command` flag
4. Replace pylint with a simple example. This eliminates any issues due to pylint.
5. Check the runfiles environment; nested binaries can interact poorly sometimes
6. I see the test changes the current working directory? If something is looking for a path relative to the execroot, that could confuse things.

The windows launcher source code is here: https://github.com/bazelbuild/bazel/tree/master/src/tools/launcher


To see how the launcher is built, see [here](https://github.com/bazelbuild/bazel/blob/master/src/main/starlark/builtins_bzl/common/python/py_executable_bazel.bzl#L318). Note that is the starlarkified code from Bazel head. The error is Bazel 6, which would be using the Java impl. The two _should_ be functionally the same, though.

aignas commented:
@rickeylev, it seems that trying to run `pylint_with_deps` inside of a `py_test` was the problem here and I have changed it to use `run_binary` instead to simulate better how the tool would be run in actual use and that has fixed the issue with the Windows test.

I have rebased on the latest master and added a `CHANGELOG` entry.

rickeylev commented on tests/entry_points/BUILD.bazel:
```diff
@@ -0,0 +1,26 @@
+py_test(
+    name = "py_console_script_gen_test",
+    srcs = ["py_console_script_gen_test.py"],
+    imports = ["../.."],
```
This imports line is so the gen_lib works, right? Please move the imports part over to the library; this makes sure the library is the one specifying how it is imported, not the binary, so that it is imported uniformly.

aignas replied:
TIL, was adding the imports out of being used to seeing them in generated `gazelle` code.

rickeylev commented on python/private/py_console_script_binary.bzl:
```diff
@@ -0,0 +1,83 @@
+    Args:
+        name: str, The name of the resulting target.
+        pkg: The package for which to generate the script.
+        entry_points_txt: The file to be used for parsing the available
+            console_script values. Defaults to searching for `entry_points.txt`
+            in the `dist_info` filegroup in the same package as the `pkg`
+            Label.
```
```suggestion
        entry_points_txt: optional target, the entry_points.txt file to parse
            for available console_script values. It may be a single file, or
            group of files, but must contain a file named `entry_point.txt`.
            If not specified, defaults to the `dist_info` target in the same
            package as the `pkg` Label.
```

aignas replied:
Thanks, better wording than mine.

rickeylev commented on python/private/py_console_script_binary.bzl:
```diff
@@ -0,0 +1,83 @@
+def _dist_info(pkg):
+    """Return the first candidate for the dist_info target label.
+
+    We cannot do `Label(pkg)` here because the string will be evaluated within
```
Similarly, if a user passes Label() (because they need for force resolution within their context for bzlmod), won't this code fail? Because it's expecting a str arg, not a Label object.

NBD for now. Just put a TODO and we can fix in a follow up PR

aignas replied:
Creating a ticket and leaving a TODO was more time than fixing the case to handle if the user passes the `Label` imho, so included the fix here.

rickeylev commented on python/private/py_console_script_gen.py:
```diff
@@ -0,0 +1,176 @@
+    return re.sub(r"[-_.]+", "_", name).lower()
+
+
+def _guess_entry_point(guess: str, console_scripts: dict[string, string]) -> str | None:
```
I'm -1 on this normalized guessing; leaving a comment to flag this; let's resolve on the issue.

aignas replied:
Removed normalizing.

rickeylev commented on CHANGELOG.md:
```diff
@@ -17,6 +17,21 @@ A brief description of the categories of changes:
+
+### Added
+
+* (bzlmod, entry_point) Introduce a new
+  [`py_console_script_binary`](./docs/py_console_script_binary.md) macro to
+  `rules_python` workspace to produce a `py_binary` target for a
+  `console_script` entry within `entry_points.txt` `dist-info` file from a
+  Python wheel.
```
```suggestion
* (bzlmod, entry_point) Added
  [`py_console_script_binary`](./docs/py_console_script_binary.md), which
  allows adding custom dependencies to a package's entry points and customizing
  the `py_binary` rule used to build it.
```

rickeylev reviewed (changes requested):
Alrighty, almost there!

aignas commented on examples/bzlmod/entry_points/BUILD.bazel:
```diff
@@ -0,0 +1,95 @@
+    ],
+)
+
+# Here we create a file that has some pylint errors detected only with the
```
Thanks, whilst updating the docs I wanted to have a rule that would allow me to take code snippets from the `examples/entry_points/BUILD.bazel` and your suggestion of moving all of the test code to a different `BUILD.bazel` makes a lot of sense at least in a sense that the comments directed towards maintainers will be next to the tests and the comments directed to readers of the examples will be next to the target definitions.

Moving them to `//entry_points/tests` along with the test files.

aignas commented on examples/bzlmod/entry_points/pylint_deps_test.py:
```diff
@@ -0,0 +1,67 @@
+        entry_point = pathlib.Path(runfiles.Create().Rlocation(rlocation_path))
+        self.assertTrue(entry_point.exists(), f"'{entry_point}' does not exist")
+
+        proc = subprocess.run(
```
Added a comment as to why `pylint_with_deps --version` works but `pylint_with_deps file.py` doesn't.

rickeylev commented on python/private/py_console_script_binary.bzl:
```diff
@@ -32,8 +32,9 @@ def _dist_info(pkg):
+        * Label("@pypi//pylint:pkg")
     """
-    return pkg.replace(":pkg", "") + ":dist_info"
+    return str(pkg).replace(":pkg", "") + ":dist_info"
```
```suggestion
    # str() is called to convert Label objects
    return str(pkg).replace(":pkg", "") + ":dist_info"
```

rickeylev commented on python/private/py_console_script_binary.bzl:
```diff
@@ -0,0 +1,86 @@
+        * @pypi//pylint:pkg
+        * Label("@pypi//pylint:pkg")
+    """
+    # str() is called to convert Label objects
```
```suggestion

    # str() is called to convert Label objects
```
